IdentityCore/src/webview/background/MSIDBackgroundTaskManager.m

```diff
@@ -0,0 +1,137 @@
+//------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+#import "MSIDBackgroundTaskManager.h"
+#import "MSIDAppExtensionUtil.h"
+
+@interface MSIDBackgroundTaskManager()
+
+@property (nonatomic) NSMutableDictionary *taskDictionary;
+
+@end
+
+@implementation MSIDBackgroundTaskManager
+
+#pragma mark - Init
+
+- (id)initInternal
+{
+    self = [super init];
+    if (self)
+    {
+        _taskDictionary = [NSMutableDictionary new];
+    }
+    return self;
+}
+
++ (MSIDBackgroundTaskManager *)sharedInstance
+{
+    static dispatch_once_t once;
+    static MSIDBackgroundTaskManager *singleton = nil;
+    
+    dispatch_once(&once, ^{
+        singleton = [[MSIDBackgroundTaskManager alloc] initInternal];
+    });
+    
+    return singleton;
+}
+
+#pragma mark - Implementation
+
+/*
+ Background task execution:
+ https://forums.developer.apple.com/message/253232#253232
+ */
+
+- (void)startOperationWithType:(MSIDBackgroundTaskType)type
+{
+    if (![NSThread isMainThread])
```

Comment:
Why does it need to be run on the main thread?

Reply:
To avoid main thread checker warnings because of accessing UIKit APIs from bg thread.

Reply:
ah, ok :)

Reply:
Is it actually logs a warning? This is from the doc of `beginBackgroundTaskWithName:expirationHandler`:
"This method can be safely called on a non-main thread."

Reply:
Good point. I assumed this behaves similarly to others UIApplication methods. Will fix.